### DISCUSSION TITLE
Missing citation / reference from time to time

### DISCUSSION THREAD

**Discussion by florentsbd** [microsoft/sample-app-aoai-chatGPT#668] [General]
Hi Everyone,

Firstly, This web app works great most of the time !
However, from to time, I see that the citation / reference are missing despite the fact that the answer is going from our indexed content (which I was able to figure out asking follow up questions).

![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/150296228/ab9855b6-095d-4968-a83c-cea8e2563503)

Another example :
![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/150296228/955dec05-363f-4c5e-a5a5-5b8bca5c2ac0)

Have you experienced the same ?

When using Azure AI Studio, I have no problem :
![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/150296228/b2076691-e7f1-4a6c-bbd5-9b54c95ae241)

Thanks.

Regards,
Florent

**Comment by florentsbd** [comment]
I actually figured out this issue.
This web app is doing a all-in-one call to Azure OpenAI using extension with extra parameters for the search query.
See here [https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#completions-extensions](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#completions-extensions).
That Azure Open AI extension has a bug and doesn't always include references (adding [docx] tag) in the response which lead to missing citation (AnswerParser.tsx needs some [docx] tags to display those citations).
So conclusion, it is not a bug in the web app but in the Azure Open AI Extension API endpoint.
I fixed that issue by doing 2 different calls changing the logic in the app.py where I first do a call to the Azure AI Search index to fetch all relevant documents and then do a call to Azure Open AI / LLM to get my completion augmenting the prompt with relevant retrieved document. I then make sure my completion include all the [docx] tags. That way it works perfectly and I have much better control on what is retrieved from the search too.
Hope it will help others.

**Comment by reminegrier** [comment]
Hi Florent, 

Thank you for posting your solution. I'd like to implement this logic (by 2 steps) to enhance all the process and keep control at each step.

Would it be possible, please, to share with us the exact way you made it by sharing your app.py ?

I really would appreciate to see the best way about implementing this solution,

Regards
